### PR TITLE
Remove enumerable from ERC721 and add an ERC721Enumerable extension

### PR DESCRIPTION
**Pull request by Amxx**
EnumerableSet and EnumerableMap heavily rely on sstore & sload, which are very expensive operations. At the same time, the Enumerable aspect can be tracked offchain using tools like thegraph.

Developer should have the choice between having this enumerable functionality (and pay the gas price), or drop it if onchain access to these information is sufficient.

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

**Comment by Amxx** [comment]
Average gas usage given by eth-gas-reporter

| Version | mint | transferFrom | burn |
|-|-|-|-|
| master:erc721 | 156250 | 94901 | 49499 |
| this:erc721 | 59397 | 53281 | 25981 |
| this:erc721enumerable | 134981 | 63952 | 41720 |

File: contracts/token/ERC721/ERC721.sol [hunk 1]
```diff
@@ -5,47 +5,36 @@ pragma solidity ^0.8.0;
...
 import "../../introspection/ERC165.sol";
 import "../../utils/Address.sol";
-import "../../utils/EnumerableSet.sol";
-import "../../utils/EnumerableMap.sol";
 import "../../utils/Strings.sol";
 
 /**
- * @title ERC721 Non-Fungible Token Standard basic implementation
+ * @title ERC721 Non-Fungible Token Standard enumerable implementation
  * @dev see https://eips.ethereum.org/EIPS/eip-721
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol:15]:
```suggestion
 * @dev Implementation of https://eips.ethereum.org/EIPS/eip-721[ERC721] Non-Fungible Token Standard, including
 * the Metadata extension, but not including the Enumerable extension, which is available separately as
 * {ERC721Enumerable}.
```

**Review by frangio — COMMENTED** [review]
Looks good overall.

File: contracts/token/ERC721/ERC721Enumerable.sol [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        if (from == address(0)) {
+            _addTokenToAllTokensEnumeration(tokenId);
+        } else if (from != to) {
+            _removeTokenFromOwnerEnumeration(from, tokenId);
+        }
+        if (to == address(0)) {
+            _removeTokenFromAllTokensEnumeration(tokenId);
+        } else if (to != from) {
+            _addTokenToOwnerEnumeration(to, tokenId);
+        }
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/token/ERC721/ERC721Enumerable.sol:79]: what about with `else if` ?

File: test/token/ERC721/ERC721.behavior.js [hunk 3]
```diff
@@ -0,0 +1,971 @@
...
+
+      it('reverts when queried for non existent token id', async function () {
+        await expectRevert(
+          this.token.tokenURI(nonExistentTokenId), 'ERC721Metadata: URI query for nonexistent token',
+        );
+      });
+
+      /**
+       * Disabled - Metadata does not use storage by default
+       */
```
[the remark this answers is not in the document]
Comment (Amxx) [reply test/token/ERC721/ERC721.behavior.js:914]: This is not part of the default behaviour, so I'll move it outside of ERC721.behavior.js, and into the test file directly. What do you think ?

[the remark this answers is not in the document]
Comment (Amxx) [reply test/token/ERC721/ERC721.behavior.js:914]: also, to do these test, we need a setTokenUri, otherwise we have to trim the test down drastically.

**Review by frangio — APPROVED** [review]
This is good to merge.

**Pull request merged**